lib/deimos/config/configurable.rb

```diff
@@ -0,0 +1,258 @@
+# frozen_string_literal: true
+
+require 'active_support/concern'
+
+module Deimos
+  # Module to allow configuration. Loosely based off of the dry-configuration
+  # gem but with several advantages:
+  # - Works with Ruby 2.3.
+  # - More succinct syntax using method_missing so you do not need to write
+  #   "config.whatever" and can just write "whatever".
+  # - Use nested blocks:
+  #   Deimos.configure do
+  #     config.kafka.ssl do
+  #       enabled true
+  #       ca_cert_file 'my_file'
+  #     end
+  #   end
+  # - Allows for arrays of configurations:
+  #   Deimos.configure do |config|
+  #     config.producer do
+  #       class_name 'MyProducer'
+  #       topic 'MyTopic'
+  #     end
+  #   end
+  # - Allows to call `configure` multiple times without crashing.
+  # - Allows to lazy-set default values by passing a proc as a default:
+  #  Deimos.configure do |config|
+  #   setting :my_val, default_proc: proc { MyDefault.calculated_value }
+  #  end
+  # - Support for setting up and automatically calling deprecated configurations.
+  # - Support for configuration callbacks.
+  module Configurable
+    extend ActiveSupport::Concern
+
+    ConfigSetting = Struct.new(:value, :default_value, :default_proc, :deprecation) do
+
+      # Reset value back to default.
+      def reset!
+        if self.value.is_a?(ConfigStruct)
+          self.value.reset!
+        elsif self.default_proc
+          self.value = self.default_proc.call
+        else
+          self.value = self.default_value
+        end
+      end
+
+      # :nodoc:
+      def clone
+        setting = ConfigSetting.new(self.value, self.default_value,
+                                    self.default_proc, self.deprecation)
+        setting.reset!
```

Comment:
Just for clarity, why does cloning reset the new object?

Reply:
Hmm - you're right, this shouldn't be a clone method. I'll rename it.